### PR TITLE
dashlane-cli 1.5.0 (new formula)

### PR DESCRIPTION
**Pull request by Mikescops**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Reopening https://github.com/Homebrew/homebrew-core/pull/124427 as repository has now enough reputation.

Publishing dashlane-cli (https://github.com/Dashlane/dashlane-cli) to homebrew.

### CODE REVIEW

File: Formula/dashlane-cli.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    strategy :github_latest
+  end
+
+  depends_on "node@16"
+
+  def install
+    Language::Node.setup_npm_environment
+    platform = OS.linux? ? "linux" : "macos"
+    libc = OS.linux? ? "glibc" : "unknown"
+    arch = `uname -m`
```
Comment (SMillerDev) [inline Formula/dashlane-cli.rb:21]:
```suggestion
    arch = Hardware::CPU.arch.to_s
```

File: Formula/dashlane-cli.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    system "npm", "ci"
+    system "npm", "run", "build"
+    system "./prepare-pkg.sh", libc
+    system "npx", "pkg", ".", "-t", "node16-#{platform}-#{arch.chomp}", "-o", "dcli"
+    bin.install "dcli"
```
Comment (SMillerDev) [inline Formula/dashlane-cli.rb:26]:
```suggestion
    system "npx", "pkg", ".", "-t", "node16-#{platform}-#{arch.chomp}", "-o", bin/"dcli"
    bin.install "dcli"
```

File: Formula/dashlane-cli.rb (hunk 1 above)
Comment (Mikescops) [reply Formula/dashlane-cli.rb]: Thanks for the suggestion!

**Comment by ZhongRuoyu** [comment]
https://www.githubstatus.com/incidents/bypf7bqdh4g0

**Comment by SMillerDev** [comment]
If upstream supports it, so should we.

**Comment by Mikescops** [comment]
> If upstream supports it, so should we.

@SMillerDev so I made tests with a real ubuntu machine and everything works well with linuxbrew, but on the runner VM it doesn't work, do you have any hints on why it could fail?

**Comment by SMillerDev** [comment]
```
==> npx pkg . -t node16-macos-arm64 -o bin/dcli
> pkg@5.8.0
> Fetching base Node.js binaries to PKG_CACHE_PATH
```
It MUST not download it's own node binaries

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate: fatal error: file not in an order that can be processed (link edit information does not fill the __LINKEDIT segment): /private/tmp/dashlane-cli-20230303-20391-an0qba/dashlane-cli-1.3.0/.brew_home/.pkg-cache/v3.4/fetched-v16.16.0-macos-arm64-signed
/private/tmp/dashlane-cli-20230303-20391-an0qba/dashlane-cli-1.3.0/.brew_home/.pkg-cache/v3.4/fetched-v16.16.0-macos-arm64-signed: the codesign_allocate helper tool cannot be found or used
> Error! Cannot generate bytecode
  pkg fails to run "codesign" utility. Due to the mandatory signing
  requirement of macOS, executables must be signed. Please ensure the
  utility is installed and properly configured.
```

It shouldn't try and sign it's own binaries since the certificate for that is probably not distributed

**Comment by Mikescops** [comment]
> It MUST not download it's own node binaries

It's not downloading anything but reusing the local nodejs install.

> It shouldn't try and sign it's own binaries 

If they are not signed the kernel will reject the binaries

**Comment by Mikescops** [comment]
> > If they are not signed the kernel will reject the binaries
> 
> But it shouldn't be signed by the install script because that would require a public certificate, which means everyone can sign with it.

We're doing an "ad-hoc" signature, there is no cryptographic proof (CMS) here, it's quite similar as not being signed but it at least allows anyone to self-sign their build (from source) and let the binary be allowed by the kernel (and for older macos version (<11) access to the keychain). You can read more [here](https://developer.apple.com/documentation/security/seccodesignatureflags/kseccodesignatureadhoc)

Concerning the fact that signing is mandatory on 11+, see this: https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-universal-apps-release-notes

File: Formula/dashlane-cli.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  on_macos do
+    # macos requires binaries to do codesign
+    depends_on xcode: :build
+    # macos 12+ only
+    depends_on macos: :monterey
+  end
+
+  on_linux do
+    depends_on "glibc" if Formula["glibc"].any_version_installed?
```
Comment (Mikescops) [inline Formula/dashlane-cli.rb:26]: just a clueless attempt to make the linux runner work

**Comment by Mikescops** [comment]
I investigated the issue more, here is what I did.

I pulled the container image used by the CI and followed all the step done by it.

What is interesting to note is that when installing from the file, the resulting binary works well.

Then the CI creates a bottle and install it.

After installing from the bottle we get the same weird error as on the CI.

```
linuxbrew@06596aa47cce:~/homebrew-core$ brew test --verbose dashlane-cli
==> Testing dashlane-cli
==> /home/linuxbrew/.linuxbrew/Cellar/dashlane-cli/1.4.0/bin/dcli --version
pkg/prelude/bootstrap.js:1
z��*!#�ڳ�ӼG��ϯ
              5~)���*�f�V       e򁍃�ol����V�L�5Ѷ��*�m�▒�A�#]a
^

SyntaxError: Invalid or unexpected token
    at new Script (node:vm:101:7)
    at readPrelude (node:internal/bootstrap/pkg:29:13)
    at node:internal/bootstrap/pkg:36:18
    at node:internal/bootstrap/pkg:43:4
    at node:internal/bootstrap/pkg:44:2
```

So what I did is to tar -xzvf the bottle and take a look at the binary. And spoiler it works well.
Which means something wrong happens during the install from bottle.

To compare the binaries I just started to use the command `file` on it:

```
linuxbrew@06596aa47cce:~/homebrew-core$ file /home/linuxbrew/.linuxbrew/Cellar/dashlane-cli/1.4.0/bin/dcli
/home/linuxbrew/.linuxbrew/Cellar/dashlane-cli/1.4.0/bin/dcli: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /home/linuxbrew/.linuxbrew/lib/ld.so, BuildID[sha1]=06d669b4b84f03ea5198d36584f920d542ca0ab9, for GNU/Linux 2.6.32, stripped
```

```
dashlane-cli/1.4.0/bin/dcli: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=06d669b4b84f03ea5198d36584f920d542ca0ab9, for GNU/Linux 2.6.32, stripped
```

As you can after being installed with the command `brew install ./dashlane-cli--1.4.0.x86_64_linux.bottle.tar.gz` the interpreter has been changed and is indeed unable to interpret the binary.

@SMillerDev is it something you have experienced before and does it qualify for a bug report?

File: Formula/dashlane-cli.rb (hunk 3 above)
Comment (Mikescops) [reply Formula/dashlane-cli.rb:26]: didn't change anything, also see my investigation below, the problem is rather on the interpreter

**Comment by SMillerDev** [comment]
> We're doing an "ad-hoc" signature, there is no cryptographic proof (CMS) here, it's quite similar as not being signed but it at least allows anyone to self-sign their build (from source) 

But building from source is not a supported configuration and ad-hoc signatures don't work on other devices as far as I know so it would fail as soon as anyone installs it outside that one CI server.

It's much better to let brew handle this.

**Comment by Mikescops** [comment]
Concerning the signature, I've open a PR on our packaging dependency to allow not signing the package on macos https://github.com/vercel/pkg/pull/1878

On the topic of the interpreter, you haven't comment @SMillerDev, any advice?

**Comment by SMillerDev** [comment]
No clue, it's the first time I see this happen.

**Pull request closed without merging**